Ability to show exported variables only when viewed within their scene

Meorge opened the pull request:
This PR adds a new annotation to GDScript, `@export_if_root`, which when placed after an `@export` annotation, causes the annotated property to only show in the inspector if the node is the root of the current scene. This is useful for hiding irrelevant information when editing nested scenes.

There is a discussion thread at https://github.com/godotengine/godot-proposals/discussions/10017 and a formal proposal at https://github.com/godotengine/godot-proposals/issues/10024. More details about the reasoning behind the feature are provided there (especially in the proposal). Before merging anything, I think it may be wise for us to refine the syntax.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/10024*

Meorge committed:
@export_if_root annotation added and seems to be working!

theraot commented:
This is technically not a review, but I can tell you right now that you should add documentation for `PROPERTY_USAGE_SHOW_IF_ROOT`. But perhaps you want to wait until you iron out the other details. I'll also make a post in the proposal shortly.

AThousandShips commented:
There's some formatting errors, please run [formatting](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html) and [add to documentation](https://docs.godotengine.org/en/latest/contributing/documentation/updating_the_class_reference.html) before we can run CI on this

Meorge commented:
I can certainly do that! For the documentation specifically, should I go ahead and write up the documentation for the implementation as it is currently, even if the details (such as the name of the annotation) are likely to change before being merged in?

AThousandShips commented:
You should at least generate them, but you can leave a temporary description until things are decided on

Meorge commented:
Alrighty, thanks! Will try to get those done today then 😄

Meorge committed:
Ran clang-format on modified files and improved comments a bit

Meorge committed:
Improved error message for @export_if_root

With the current implementation, @export_if_root must come after an @export annotation since it modifies the property's export_info

Meorge committed:
Documentation added

@export_if_root and PROPERTY_USAGE_SHOW_IF_ROOT have descriptions in their respective locations in the documentation

AThousandShips commented on core/core_constants.cpp:
```diff
@@ -710,7 +710,7 @@ void register_global_constants() {
 	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_SECRET);
-
+	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_SHOW_IF_ROOT);
 	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_DEFAULT);
```
```suggestion

	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_DEFAULT);
```
Keep an empty line separating these for readability

AThousandShips commented on core/object/object.h:
```diff
@@ -120,7 +120,7 @@ enum PropertyUsageFlags {
 	PROPERTY_USAGE_SECRET = 1 << 29, // Export preset credentials that should be stored separately from the rest of the export config.
-
+	PROPERTY_USAGE_SHOW_IF_ROOT = 1 << 30, // Only show a property in the inspector if the node is the current root.
 	PROPERTY_USAGE_DEFAULT = PROPERTY_USAGE_STORAGE | PROPERTY_USAGE_EDITOR,
```
```suggestion

	PROPERTY_USAGE_DEFAULT = PROPERTY_USAGE_STORAGE | PROPERTY_USAGE_EDITOR,
```

AThousandShips commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -577,6 +577,17 @@
+		<annotation name="@export_if_root">
+			<return type="void" />
+			<description>
+				When placed between an [code]@export[/code] annotation and the exported property, causes the exported property to only be displayed in the Inspector dock if the currently-selected node is the root node of the current scene.
```
```suggestion
				When placed between an [annotation @export] annotation and the exported property, causes the exported property to only be displayed in the Inspector dock if the currently-selected node is the root node of the current scene.
```

Meorge committed:
Apply suggestions from code review

Co-authored-by: A Thousand Ships (she/her) <96648715+AThousandShips@users.noreply.github.com>

AThousandShips commented:
I feel this annotation is far too specific, I can think of few cases when I'd want something to show up *only* when the node is the root

Honestly I'd personally be more interested in when it *isn't* the root but instead is in an instanced scene, that makes far more sense to me

But I feel this is far too specific and also far too easy to work around in custom code to warrant a custom annotation, you can simply do this with `_validate_property` and `EditorInterface.get_edited_scene_root()` (though that might not work easily in exported projects, but that's something to solve with other features like editor only compiled code, not something against that workaround)

AThousandShips commented:
> I'm not sure at which point it's acceptable to take a proposal and try to implement it

There really isn't any limit for that, the only thing to consider is that you might be wasting time if you try to implement a proposal before it's gotten good support and ends up rejected because it's not in high enough demand, or having to redesign it multiple times because the discussion changes and other details are focused on

So beyond that letting a proposal sit until some stronger consensus has been reached it's up to you, but jumping on a proposal before it's been discussed deeply enough can risk making a PR when the discussion would otherwise have led to dismissing the idea or some workaround being found or some different solution being decided on

Meorge commented:
Closing this as there's not a lot of interest in it, and/or if the `private` access modifier ends up going forward, a similar function could potentially be a part of that.
